### DISCUSSION TITLE
Need to be able to wait for a token to resolve before recoil initialization

### DISCUSSION THREAD

**Discussion by Cpt-Falcon** [facebookexperimental/Recoil#2068] [Q&A]
Before recoil begins to initialize atoms and selectors, which in my case will make API calls that requires auth/token, I need to await a promise containing the token.

My original try was to use the initialize state in RecoilRoot. Unfortunately, recoil root initialize state  doesn't work with an async function. The initialization of atoms and selectors proceeds before the token and initialize state is fully resolved, and therefore fails. I would like to see recoil root initialize state block until any async await operations were finished resolving.

My alternative approach was to use an atom with a default selector initializer like this.

```
export const connectionService = atom<IConnectionService>({
    key: 'connectionService',
    default: selector({
        key: 'connectionServiceInit',
        get: async () => {
            const shell = useShellAPI();
            const token = await shell.getToken();
            const connect = new ConnectionService(token as string);
            return connect;
        },
        dangerouslyAllowMutability: true,
    }),
    dangerouslyAllowMutability: true,
});
```

This is not as good IMO. I had to use dangerous mutability here because the connection service changes itself under the hood, like updating timeout fields and whatnot. However, all dependent selectors will wait for the connection service to finishing resolving so this works pretty darn well.

By the way, top level async await does not work because I'm using NX with federated modules. 

So I'm open to other options but preferably to level initialize state in recoil root would take in an async function and wait for it to finish before doing anything else

**Comment by drarmstr** [comment]
If you really want you could make a wrapper around `<RecoilRoot>` to handle the async initialization.  Though, that's a pretty big hammer.  Having atoms/selectors that require the async token rely on an atom or selector that contains the token would hopefully capture the granularity that you need.  It looks like your proposed approach mostly dos this.  Note that you cannot use hooks inside selector evaluation due to the rules of React hooks.  To avoid the mutability issue you can have the atom just contain some token or handle of the connection vs the actual stateful mutable connection service object itself.  You also don't need a default selector but can set the atom default to a `Promise` that resolves to the token or handle when ready.
